### ISSUE TITLE
numpy.all support

### ISSUE THREAD

**Issue by jamescasbon** [Parquery/icontract#97]
I'm trying to work out how to add numpy contracts without over repetition.  Most numpy functions return vectors and need a np.all to return a bool.

For example, checking input is positive:
```
>>> import numpy as np                                                                                                                                                                      
>>> x = np.arange(-3, 4)                                                                                                                                                                    
>>> x                                                                                                                                                                                       
array([-3, -2, -1,  0,  1,  2,  3])
>>> x > 0                                                                                                                                                                                   
array([False, False, False, False,  True,  True,  True])
>>> np.all(x > 0)                                                                                                                                                                           
False
```
I  can of course add a precondition with  

    @require(lambda x: np.all(x > 0))

But this gets tedious and verbose (the lambda and the all).  

I tried to extend require to a version like vrequire (vector require) which would allow me to: 

    @vrequire(lambda x: x> 0)

but its not easy as I have to define a condition with the right argument name.  Any tips on how I might do this?

thanks!

**Comment by jamescasbon** [comment]
Ah, it's actually worse than that. 

```
from icontract import require
import numpy as np

@require(lambda x: np.all(x > 0))
def frobulate(x: np.ndarray):
    return True

frobulate(np.array([-3,3]))
```

```
Traceback (most recent call last):
  File "test/foo.py", line 8, in <module>
    frobulate(np.array([-3,3]))
  File "/home/james/.conda/envs/tge/lib/python3.6/site-packages/icontract/_checkers.py", line 285, in wrapper
    _assert_precondition(contract=contract, resolved_kwargs=resolved_kwargs)
  File "/home/james/.conda/envs/tge/lib/python3.6/site-packages/icontract/_checkers.py", line 110, in _assert_precondition
    msg = icontract._represent.generate_message(contract=contract, condition_kwargs=condition_kwargs)
  File "/home/james/.conda/envs/tge/lib/python3.6/site-packages/icontract/_represent.py", line 387, in generate_message
    a_repr=contract._a_repr)
  File "/home/james/.conda/envs/tge/lib/python3.6/site-packages/icontract/_represent.py", line 334, in repr_values
    recompute_visitor.visit(node=lambda_inspection.node.body)
  File "/home/james/.conda/envs/tge/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/home/james/.conda/envs/tge/lib/python3.6/site-packages/icontract/_recompute.py", line 257, in visit_Call
    args.append(self.visit(node=arg_node))
  File "/home/james/.conda/envs/tge/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/home/james/.conda/envs/tge/lib/python3.6/site-packages/icontract/_recompute.py", line 239, in visit_Compare
    if not comparison:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

**Comment by mristin** [comment]
Hi @jamescasbon ,
Thanks a lot for reporting the problem! Let us first fix the value error and then we can have a look at how you can introduce shorter versions of require (`vrequire`) into your code base.

The following snippet reproduced the problem with the numpy and comparisons:

```python
#!/usr/bin/env python3

import numpy as np

def main():
    x = np.array([-3, 3])
    y = x > 0
    z = not y

if __name__ == '__main__':
    main()
```
which gives:
```
Traceback (most recent call last):
...
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

In icontract I used `not` during the recomputation to produce the error message (so that the user can easier see what causes the breach of the contract). In https://github.com/Parquery/icontract/blob/b0f1f4ad462ccec5db6c8c924185874a195a614b/icontract/_recompute.py#L239:

```python3
            if not comparison:
                result = False
                break
```

which raises the ValueError you reported.

I'm having a look at the code right now and trying to fix this issue.

**Comment by mristin** [comment]
Hi @jamescasbon ,
I fixed the problem with ValueError, bumped the version to 2.0.3 and published it on pypi. Could you please check if it still persists with icontract 2.0.3?

Once you confirm me that the problem is resolved, we'll have a look at how to simplify the preconditions.

**Comment by jamescasbon** [comment]
Wow, amazing. Thanks for the fix!

Yeah, just thinking it through perhaps we dont *need* a vrequire as a short alias to all can work:

```
from icontract import require, ensure
import numpy as np
from numpy import all as A

@ensure(lambda x: A(x > 0))
def frobulate(x: np.ndarray):
    return x

frobulate(np.array([-3, 3]))
```

But one thing that slightly bothers me is the lack of a 'point free style'.  Lets say I define a complex function

     def my_complex_condition(x): return np.all(np.isfinite(x) & (np.abs(x) < 100))

Then I can't just write:

    @require(my_complex_condition) 

as we need to rebind the argument name if the function takes a y

```
@require(lambda y: my_complex_condition(y))
def fn(y):
    ....
````

It's not your fault that python's lambda syntax is so verbose.  I wonder if I'm missing any shorter syntax? something like this springs to mind:

    @require({'x': my_complex_condition})
    def fn(x, y): 

I think they should have made λ an alternate spelling of lambda !

**Comment by mristin** [comment]
You could try to have a factory methods for the lambda somehow, but my gut feeling is that any such approach would be slow. We had some discussion about it o the issue https://github.com/Parquery/icontract/issues/61.

Maybe you can ask on python-ideas mail list that λ be included in Python 3.8?

It indeed reads much cleaner:
```python
@require(λ y: my_complex_condition(x=y))
...
```

Passing the dictionary is an option, but that would require more work. At the moment, I'm short on time so I wouldn't be able to make it. If you have some time, you can go ahead and make a pull request. I can definitely help you if you have any questions regarding the existing code.

**Comment by jamescasbon** [comment]
OK, I thnk I have enough for now !
